### ISSUE TITLE
Logging has more overhead than expected

### ISSUE THREAD

**Issue by ad8e** [clearml/clearml#1378]
## Describe the bug
Wandb takes 0.2 ms to log each step. ClearML takes 1 ms (5x as much), and also spikes to 100 ms every 14-15 steps. This is wandb's logging:
```
wandb.log(data=log_dict, step=counter, commit=True)
```
This is ClearML's corresponding logging:
```
logger = clearml.Task.current_task().get_logger()
logger: clearml.Logger
for name, v in log_dict.items():
    logger.report_scalar(title=name, series=name, value=v, iteration=counter)
```

Wandb's times:
```
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0001 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0001 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0001 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
Time: 0.0002 seconds
```

ClearML's times:
```
Time: 0.1012 seconds
Time: 0.0008 seconds
Time: 0.0008 seconds
Time: 0.0008 seconds
Time: 0.0008 seconds
Time: 0.0010 seconds
Time: 0.0010 seconds
Time: 0.0011 seconds
Time: 0.0009 seconds
Time: 0.0009 seconds
Time: 0.0009 seconds
Time: 0.0008 seconds
Time: 0.0010 seconds
Time: 0.0010 seconds
Time: 0.1014 seconds
Time: 0.0009 seconds
Time: 0.0010 seconds
Time: 0.0009 seconds
Time: 0.0008 seconds
Time: 0.0008 seconds
Time: 0.0009 seconds
Time: 0.0011 seconds
Time: 0.0010 seconds
```

Measurement uses this context manager wrapped around the log call:
```
# https://stackoverflow.com/questions/33987060/python-context-manager-that-measures-time
class catchtime:
    def __enter__(self):
        self.start = monotonic()
        return self

    def __exit__(self, type, value, traceback):
        print(f"Time: {monotonic() - self.start:.4f} seconds")
```

## Expected behaviour
I expect logging to be much cheaper. That spike every 14-ish steps worries me a lot as well.

## Environment
* app.clear.ml
* ClearML SDK Version 1.17.1
* Python 3.10
* Linux

Slack discussion: https://clearml.slack.com/archives/CTK20V944/p1740274510594309

**Comment by ad8e** [comment]
If I increase the points per step to 50-ish, ClearML becomes unusable:
```
Time: 0.0065 seconds
Time: 0.1066 seconds
Time: 0.1066 seconds
Time: 0.0063 seconds
Time: 0.1068 seconds
Time: 0.1069 seconds
Time: 0.1067 seconds
Time: 0.0062 seconds
Time: 0.1078 seconds
Time: 0.1068 seconds
Time: 0.0064 seconds
Time: 0.1067 seconds
Time: 0.1066 seconds
Time: 0.1066 seconds
Time: 0.0064 seconds
```
ClearML's logging takes 3 times longer than the entire forward/backward/optimizer.

Wandb still stays performant in this case:
```
Time: 0.0004 seconds
Time: 0.0004 seconds
Time: 0.0004 seconds
Time: 0.0004 seconds
Time: 0.0004 seconds
```

**Comment by ad8e** [comment]
When I add an extra `print("flushing")` to `def add_event(self, ev):` in `reporter.py`, the time spikes go away:

```
        if self._queue_size >= self._flush_threshold:
            print("flushing")
            self.flush()
```

```
Time: 0.0069 seconds
Time: 0.0067 seconds
flushing
Time: 0.0069 seconds
flushing
Time: 0.0069 seconds
flushing
Time: 0.0069 seconds
Time: 0.0068 seconds
flushing
Time: 0.0070 seconds
flushing
Time: 0.0070 seconds
Time: 0.0066 seconds
```
Overall step time has not increased. So this extra print is not just shuffling the GC issue to a different part of the code, it's actually mitigated it.

**Comment by ad8e** [comment]
```
def flush(self):
    while isinstance(self._queue, PrQueue) and self._queue.is_pending():
        sleep(0.1)
```
This explains the 0.1 sec delays. It's not from Python GC, it's ClearML's code. These delays block training synchronously, which is very expensive in large scale runs.

**Comment by ad8e** [comment]
I also bisected the overhead to these lines:
https://github.com/clearml/clearml/blob/342e1b35f8be532acdc27d74402482a4d67a19cf/clearml/backend_interface/metrics/reporter.py#L379 costs 0.5 ms
https://github.com/clearml/clearml/blob/342e1b35f8be532acdc27d74402482a4d67a19cf/clearml/backend_interface/metrics/reporter.py#L138 costs 2.4 ms
https://github.com/clearml/clearml/blob/342e1b35f8be532acdc27d74402482a4d67a19cf/clearml/backend_interface/metrics/reporter.py#L160 costs 0.4 ms
https://github.com/clearml/clearml/blob/342e1b35f8be532acdc27d74402482a4d67a19cf/clearml/backend_interface/metrics/reporter.py#L163 costs 0.12 ms, but sometimes 1000.12 ms

My method was to gradually uncomment lines, then see the difference in overhead.

This entire function chain should be looked over carefully, because there are lines like https://github.com/clearml/clearml/blob/342e1b35f8be532acdc27d74402482a4d67a19cf/clearml/logger.py#L176 which are not needed. In C++, these would be free, but they are less free in Python. And https://github.com/clearml/clearml/blob/342e1b35f8be532acdc27d74402482a4d67a19cf/clearml/logger.py#L1529, which would probably be faster with get(), so that the common case of the title existing does not require a double dictionary lookup. ScalarEvent also seems unnecessarily expensive.

A batched scalar metric logging function would likely be much more performant than this.

**Comment by jkhenning** [comment]
@ad8e this is not a spike - the SDK caches these reports and dumps them in an API call once 100 events or more are pending in the cache

**Comment by jkhenning** [comment]
@ad8e please try to config the following setting in you `clearml.conf` and see if it has an effect on your system:
```
sdk.development.report_use_subprocess = False
```

Also, you can try to set these settings in the same file:
```
sdk.development.worker {
    # Status report period in seconds
    report_period_sec: 20
    # The number of events to report
    report_event_flush_threshold: 300
    # ping to the server - check connectivity
    ping_period_sec: 30
}
```

**Comment by ad8e** [comment]
```
sdk.development.report_use_subprocess = False
```
This works. Time went from 6.5 ms to 0.75 ms, which is 9x faster. Skipping the `_fast_is_subprocess_alive` check is key. Overhead is down to 4% on this model.

I added an early return to flush(), to avoid the 0.1 sleep. (Later, I'll instead remove flush() from the scalar function, and instead check for flush once per training step.) Might leaving a lot of events in the queue cause any problems?
```
    def flush(self):
        if isinstance(self._queue, PrQueue) and self._queue.is_pending():
            return
            sleep(0.1)
```

**Comment by ad8e** [comment]
After Jake's mitigation, ClearML's overhead in a full-size transformer is 5.77 ms, which is 0.3%. This is much better. Checking some of the remaining time sources:
3.57 ms is from `ev = ScalarEvent(...)`.
0.6 ms is from the two `_convert_np_nan_inf(value)` lines inside ScalarEvent.
0.12 ms is from the repeated `time()` calls inside ScalarEvent.
I assume most of the rest is inside the heavy API handling inside`ApiServiceProxy`, but it's hard to check.
Each no-op function call costs ~0.02 ms. For example, there are extra `self._normalize_name` calls inside the `ev` creation, and a `self._start_task_if_needed()` which does nothing.
Applying the second suggestion:
```
sdk.development.worker {
    # Status report period in seconds
    report_period_sec: 20
    # The number of events to report
    report_event_flush_threshold: 300
    # ping to the server - check connectivity
    ping_period_sec: 30
}
```
might make the model slower by a tiny bit (0-0.05 ms).

My proposed `get()` for `_touch_title_series` was slower than the existing code.

I replaced `flush()`'s `sleep(0.1)` with a  `Return`.

With this, calling `flush()` after every step, instead of checking the queue size after every metric log (= multiple times per step), made the model marginally slower. I assume it's because the Python queue becomes slower with more elements.

I have not checked the queue implementation, but the majority of overhead from `ScalarEvent` seems to be caused by repeated API checks in `ApiServiceProxy`. Batching this to once-per-step should cut down ClearML's overhead by half.

**Comment by patrickdasilva-clearml** [comment]
Hi @ad8e , thanks for your detailed report! Your contribution is very valuable.

The discussion went stale for over a year and we have not found an explicit change we could to do improve the system, so I will close the issue for now. If the issue persists for you, please feel free to re-open this issue.
